Added Support for Toshiba TMPM46B

ganesh-ramachandran opened the pull request:
## Description
Add mbed support for TOSHIBA's TMPM46B board.

## Status
READY

## Migrations
NO

## Test Status
Pass: Mbed SDK automated test suite and greentea with ARM, GCC_ARM and IAR toolchains.

[M46B_GreenteaAutomated_TestResults.txt](https://github.com/ARMmbed/mbed-os/files/1655378/M46B_GreenteaAutomated_TestResults.txt)
[M46B_SingleAutomated_TestResults.txt](https://github.com/ARMmbed/mbed-os/files/1655379/M46B_SingleAutomated_TestResults.txt)

0xc0170 commented on targets/TARGET_TOSHIBA/TARGET_TMPM46B/trng_api.c:
```diff
@@ -0,0 +1,100 @@
+int trng_get_bytes(trng_t *obj, uint8_t *output, size_t length, size_t *output_length)
+{
+    uint32_t Data[length];
+    int timeout = 10;
```
how is  t his number calculated ? why 10?

ganesh-ramachandran replied:
The entropy seed generator (ESG) is a circuit that generates a 512-bit entropy seed using a ring oscillator. When the ESG core ends the operation, it outputs an interrupt. Interrupt generation can be checked by polling of the interrupt flag. Then an entropy seed can be read from 16 blocks (ESGBLK00 to 15)
Timeout 10 is used for 10 iteration to confirm the ESG operation is completed.
- Sometimes function execution completes before generating interrupts and read wrong value.
- Timeout will help to return from function, even interrupts are not generated within 10 iteration.

We will also try to change the approach.

0xc0170 replied:
@hanno-arm Please review trng addition

0xc0170 commented on targets/TARGET_TOSHIBA/TARGET_TMPM46B/sleep.c:
```diff
@@ -0,0 +1,29 @@
+
+void hal_deepsleep(void)
+{
+    hal_sleep();
```
why deepsleep = sleep here?

ganesh-ramachandran replied:
TMPM46BF10FG does not support the low power consumption mode configured with the SLEEPDEEP bit in the Cortex-M4 core.
Setting the <SLEEPDEEP> bit of the system control register is prohibited.

0xc0170 replied:
OK Thanks, if this is captured in git log or can be here in the code, so others wont be surprised?

ganesh-ramachandran replied:
Hi @0xc0170, added the reason as code comments. Thank you.

0xc0170 commented on targets/TARGET_TOSHIBA/TARGET_TMPM46B/pwmout_api.c:
```diff
@@ -0,0 +1,205 @@
+#include "tmpm46b_tmrb.h"
+
+#define MAX_COUNTER_16B         0xFFFF
+TMRB_InitTypeDef m_tmrb;
```
these 2 variables static?

ganesh-ramachandran replied:
Updated as static.

0xc0170 commented on targets/TARGET_TOSHIBA/TARGET_TMPM46B/i2c_api.c:
```diff
@@ -0,0 +1,348 @@
+    20, 24, 32, 48, 80, 144, 272, 528
+};  // SCK Divider value table
+
+I2C_clock_setting_t clk;
```
why these 2 are not `static` as well?

ganesh-ramachandran replied:
Updated as static.

0xc0170 commented:
>Can you share tests results for all 3 toolchains?

I might have missed the test results shared above. The failure MBED 11 was resolved recently, should be also green !

Travis failed, this was also fixed. Can you please rebase from master, to get both fixes in?

ganesh-ramachandran commented:
> I might have missed the test results shared above. The failure MBED 11 was resolved recently, should be also green !

We checked MBED 11 and result is OK.

> Travis failed, this was also fixed. Can you please rebase from master, to get both fixes in?

Rebased from master.

hanno-arm commented on targets/TARGET_TOSHIBA/TARGET_TMPM46B/trng_api.c:
```diff
@@ -0,0 +1,100 @@
+
+int trng_get_bytes(trng_t *obj, uint8_t *output, size_t length, size_t *output_length)
+{
+    uint32_t Data[length];
```
The `length` parameter is the maximum amount of entropy _in bytes_ that may be written to `output`, while `Data` has four times that size.

hanno-arm commented on targets/TARGET_TOSHIBA/TARGET_TMPM46B/trng_api.c:
```diff
@@ -0,0 +1,100 @@
+        // Get the calculation result
+        if(CompleteFlag == 1){
+            CompleteFlag = 0;
+            ESG_GetResult(Data); 
```
This might overflow the `Data` buffer if `length < 16`.

hanno-arm commented on targets/TARGET_TOSHIBA/TARGET_TMPM46B/trng_api.c:
```diff
@@ -0,0 +1,100 @@
+        }
+    }
+
+    *output_length = length;
```
This is not accurate: The current function at most fetches `64` bytes of entropy, and only that amount may be reported back to the user.

hanno-arm commented on targets/TARGET_TOSHIBA/TARGET_TMPM46B/trng_api.c:
```diff
@@ -0,0 +1,100 @@
+            CompleteFlag = 0;
+            ESG_GetResult(Data); 
+            for(i = 0U; i < length; i++){
+                output[i] = Data[i];
```
At most `16` bytes have been written to `Data`, but this loop reads its entire contents of user-specified size `length`, filling the remainder of `output` with non-random data.

ganesh-ramachandran commented:
Hi @hanno-arm, rewritten the entropy gathering function trng_get_bytes based on your comments.

hanno-arm commented:
Hi @ganesh-ramachandran, thank you for your rework! Could you point me to the technical reference manual for the board?

ganesh-ramachandran commented:
> Could you point me to the technical reference manual for the board?

Hi @hanno-arm, Please find the datasheet and schematics at https://os.mbed.com/platforms/TMPM46B/#technical-reference

ciarmcom commented:
ARM Internal Ref: IOTSSL-2073

0xc0170 commented:
@ganesh-ramachandran As TRNG is waiting for mbedtls approval, it might be faster to leave TRNG here and send PR adding target support separately? Or leave it as it is.

hanno-arm commented on targets/TARGET_TOSHIBA/TARGET_TMPM46B/trng_api.c:
```diff
@@ -0,0 +1,105 @@
+            ESG_SetLatchTiming(ESG_LATCH_TIMING_1);
+            // Get latchtiming value
+            latchtiming = ESG_GetLatchTiming();
+            if (Fintming >= 512U * (latchtiming + 1U) + 3U) {
```
@ganesh-ramachandran Could you explain the rationale behind this check? As far as I see, it passes if and only if `latchtiming` is zero (` == ESG_LATCH_TIMING_1`), so why not check for this directly?

ganesh-ramachandran replied:
Yes, we understood and removed the extra checking.

hanno-arm commented on targets/TARGET_TOSHIBA/TARGET_TMPM46B/trng_api.c:
```diff
@@ -0,0 +1,105 @@
+    } else {
+        // Do nothing 
+    }
+}
```
Unfortunately, the current API doesn't allow `trng_init` to have a return value to signal failure (that has been resolved in Mbed TLS 2.7.0 and will hopefully soon be merged into Mbed OS). Still, I'd like to seek for a solution which allows to distinguish initialization failure from success, in order to not have the TRNG driver potentially output non-random data because of unnoticed failure during initialization. 

Suggestion: Currently, the TRNG context `trng_s` consists of an unused `dummy` field only. This can be changed to a boolean flag indicating successful initialization, which is set on a successful call to `trng_init`, cleared on a call to `trng_free`, and checked for in `trng_get_bytes`.

ganesh-ramachandran replied:
Thank you for the suggestion. we have added the errors handling during TRNG initialization to prevent accidental output of non-random data.

hanno-arm commented on targets/TARGET_TOSHIBA/TARGET_TMPM46B/trng_api.c:
```diff
@@ -0,0 +1,105 @@
+        // wait for completing calculation
+    }
+    // check the interrupt flag 
+    if (CompleteFlag == 1) {
```
How does this work for the first call to `trng_get_bytes`? We initialize `CompleteFlag = 0`, so unless the ESG fires an interrupt at startup (does it?), this path will never be taken.

hanno-arm commented on targets/TARGET_TOSHIBA/TARGET_TMPM46B/trng_api.c:
```diff
@@ -0,0 +1,105 @@
+    *output_length = 0;
+    (void)obj;
+
+    while (TSB_ESG->ST == ESG_BUSY_SET) {
```
Why are we polling on `TSB_ESB->ST` here instead of `CompleteFlag`, which is what the ESG's IRQ handler sets?

ganesh-ramachandran replied:
TSB_ESB->ST is the status register and it indicate the process is complete or not.
While polling IRQ handler flag(CompleteFlag), interrupt will generate only once.
Hence if application will try to call two times the trng_get_bytes() function it will freeze the application. To avoid this scenario we poll TSB_ESB->ST and reconfirm IRQ handler flag (CompleteFlag) is set or not.

Now we changed to polling on IRQ handler flag(CompleteFlag), because we confirmed the upper layer file(mbed_trng.c) always calls trng_get_bytes() function in between TRNG init and TRNG free.

hanno-arm commented on targets/TARGET_TOSHIBA/TARGET_TMPM46B/trng_api.c:
```diff
@@ -0,0 +1,105 @@
+        CompleteFlag = 0;
+        // Get the calculation result
+        ESG_GetResult((uint32_t*)random); //512-bit entropy
+        for (i = 0; (i < 64) && (*output_length < length); i++) {
```
Please use `i < length` here instead of `*output_length`, and set `*output_length` only once at the end of the loop.

ganesh-ramachandran replied:
Yes, modified as per your suggestion.

0xc0170 commented:
/morph build

hanno-arm commented on targets/TARGET_TOSHIBA/TARGET_TMPM46B/trng_api.c:
```diff
@@ -0,0 +1,99 @@
+    ESG_Config();
+    // Start up the ESG core
+    ESG_Startup();
+    obj->trng_init = true;
```
Please take the return code of `ESG_Startup` into account here, and also add and obey a return code to `ESG_Config` indicating if the initialization was successful.

ganesh-ramachandran replied:
Update the code as per the suggestion.

hanno-arm commented on targets/TARGET_TOSHIBA/TARGET_TMPM46B/trng_api.c:
```diff
@@ -0,0 +1,99 @@
+
+void INTESG_IRQHandler(void)
+{
+    CompleteFlag = 1;
```
As far as I understand this code, we're continuously accumulating new entropy even if we don't need it, giving away computing time for no benefit. What about removing the interrupt handler altogether, and polling on the ESG interrupt status register directly within `trng_get_bytes`?

ganesh-ramachandran replied:
Update the code as per the suggestion.

hanno-arm commented:
@0xc0170 Could you please also request a review from @mazimkhan? Thanks!

ganesh-ramachandran commented:
Hi @0xc0170, /morph build failure resolved. Issue is with the case sensitive header file name.

hanno-arm commented on targets/TARGET_TOSHIBA/TARGET_TMPM46B/trng_api.c:
```diff
@@ -0,0 +1,88 @@
+{
+    Result ret = SUCCESS;
+    uint16_t Fintming = 560U;
+    // Confirm the ESG core stops
```
Please use the error-by-default approach used in `tmpm46b_esg.c`, initializing `ret = ERROR` and clearing it to `ret = SUCCESS` in the unique successful code path. Further, please consider the return values of `ESG_SetLatchTiming` and `ESG_SetFintiming`, too.

ganesh-ramachandran replied:
Initialized with ret = ERROR, and updated SUCCESS based on ESG_SetLatchTiming and ESG_SetFintiming return values.

hanno-arm commented on targets/TARGET_TOSHIBA/TARGET_TMPM46B/trng_api.c:
```diff
@@ -0,0 +1,85 @@
+        return -1;
+    }
+
+    while (TSB_ESG->ST == ESG_BUSY_SET)
```
@ganesh-ramachandran  As far as I see, this will only run the ESG once, but not re-request the generation of another block of entropy. According to Section 14.4 of the Technical Reference Manual, I would expect a polling on `ESG_GetIntStatus` and afterwards clearing the interrupt via `ESG_ClrInt`. Also, the manual doesn't specify whether `ESG_GetIntStatus == ENABLE` implies that `ESG_GetCalculationStatus == ESG_CALCULATION_COMPLETE`, so perhaps one should poll on `ESG_GetCalculationStatus == ESG_CALCULATION_COMPLETE` first, too.

ganesh-ramachandran replied:
> As far as I see, this will only run the ESG once, but not re-request the generation of another block of entropy.

Hi @hanno-arm, ESG hardware supports only 1 RNG instance. For regeneration of another block of entropy, ESG hardware has to stop(trng_free) and start(trng_init) again. Other comments addressed. Please confirm.

hanno-arm replied:
@ganesh-ramachandran Currently, the TRNG will only generate a single block of entropy, but the API demands that after `trng_init` has been called, the user should be able to call `trng_get_bytes` to fetch an arbitrary amount of entropy. With the previous version, this seemed to work because the TRNG's `IntStatus` was always cleared on an interrupt, which to my understanding would trigger the generation of another block of entropy. With the current code,however, this signal to generate the next block of entropy is missing.

ganesh-ramachandran replied:
Hi @hanno-arm, currently we are verifying whether the IntStatus clearing at ESG_Config will work. Please confirm.
Also please share any reference target with latest TRNG implementation.

<pre>
static Result ESG_Config(void)
{
    Result ret = ERROR;
    uint16_t Fintming = 560U;

    if (ESG_GetIntStatus() == ENABLE) {
        while(ESG_GetCalculationStatus() == ESG_CALCULATION_COMPLETE) {
           // Wait for ESG core ends the operation
        }
        ret = ESG_ClrInt(); // Clear the Interrupt
    }
    // Confirm the ESG core stops
    if (ESG_GetCalculationStatus() == ESG_CALCULATION_COMPLETE) {
        // Confirm no interrupt generation
        if (ESG_GetIntStatus() == DISABLE) {
            // Set the latch timing & output timing
            if ((ESG_SetLatchTiming(ESG_LATCH_TIMING_1) == SUCCESS) &&
                (ESG_SetFintiming(Fintming) == SUCCESS)) {
                ret = SUCCESS;
            }
        }
    }
    return ret;
}
</pre>

hanno-arm replied:
As mentioned below, the reference manual doesn't provide sufficient information for me to understand how entropy re-generation is supposed to work. 

What's the reason you modified `ESG_Config` in the latest change?

hanno-arm commented on targets/TARGET_TOSHIBA/TARGET_TMPM46B/trng_api.c:
```diff
@@ -0,0 +1,95 @@
+        return -1;
+    }
+
+    while(ESG_GetCalculationStatus() == ESG_CALCULATION_COMPLETE)
```
Do you want to poll on `ESG_GetCalculationStatus() != ESC_CALCULATION_COMPLETE` instead?

hanno-arm commented on targets/TARGET_TOSHIBA/TARGET_TMPM46B/trng_api.c:
```diff
@@ -0,0 +1,95 @@
+    }
+
+    ret = ESG_ClrInt(); // Interrupt is cleared
+    // Get the calculation result
```
I can't infer from reading the reference manual how users are supposed to request re-generation of entropy. Unless you have more detailed information on that, we are left with trial and test, here. For example, I don't know if it is necessary to re-call `ESG_Startup` after clearing the interrupt.

ganesh-ramachandran replied:
Hi @hanno-arm, we have added code for restarting ESG core at end of trng_get_bytes to generate more blocks in repeated trng_get_bytes calls.

hanno-arm commented on targets/TARGET_TOSHIBA/TARGET_TMPM46B/trng_api.c:
```diff
@@ -0,0 +1,93 @@
+       // Wait for ESG core for generating an interrupt
+    }
+
+    ret = ESG_ClrInt(); // Interrupt is cleared
```
Please return immediately if `ESG_ClrInt` fails - it's not specified what happens if you attempt to read the random block registers while the ESG is still in operation.

ganesh-ramachandran replied:
@hanno-arm, Return added if `ESG_ClrInt` fails.

> what happens if you attempt to read the random block registers while the ESG is still in operation.

When the ESG core ends the operation then only interrupt will generate.
Polling on interrupt status will ensure the core processing in completed.
After generation of interrupt, an entropy seed can be read.
ESG will not operate once the entropy is generated, hence no need to taken care while reading entropy. ESG will operate after reset and re-initialization.

hanno-arm commented on targets/TARGET_TOSHIBA/TARGET_TMPM46B/trng_api.c:
```diff
@@ -0,0 +1,93 @@
+    *output_length = i;
+    //clear and restart the ESG core
+    ESG_IPReset(); 
+    ret |= ESG_Startup();
```
As mentioned, the specification is not rich enough to judge whether this is correct. @ganesh-ramachandran What kind of tests did you run?

ganesh-ramachandran replied:
> What kind of tests did you run?

We tested with below mbedos testcases
 tests-mbedtls-multi
 tests-mbedtls-selftest

For repeated entropy generation, tested with local application and verified ESG registers are updating with new values in each call.

hanno-arm commented:
@cmonr Yes, @mazimkhan needs to review this, too.

ganesh-ramachandran commented:
> @ganesh-ramachandran As TRNG is waiting for mbedtls approval, it might be faster to leave TRNG here and send PR adding target support separately? Or leave it as it is.

@0xc0170 @hanno-arm, can we go ahead with this. Keep this PR for TRNG and create a new PR for adding target support without TRNG? Please confirm.

hanno-arm commented:
@mazimkhan Ping.

hanno-arm commented on targets/TARGET_TOSHIBA/TARGET_TMPM46B/trng_api.c:
```diff
@@ -0,0 +1,97 @@
+
+void trng_init(trng_t *obj)
+{
+    // Enable clock supply
```
Please add a check that `obj->trng_init == false` when `trng_init` is called.

mazimkhan replied:
Actually if this function is suppose to initialise the ```obj```. Then it can't trust it's contents to determine if already initialised. If it is required for the driver to know the initialised state. Then it has to save that state itself. How to do that, is implementation dependent.

hanno-arm replied:
Thanks @mazimkhan, you're absolutely right - @ganesh-ramachandran please ignore this suggestion.

mazimkhan commented on targets/TARGET_TOSHIBA/TARGET_TMPM46B/Periph_Driver/src/tmpm46b_esg.c:
```diff
@@ -0,0 +1,314 @@
+    do {
+        iprst3 = TSB_SRST->IPRST & SRST_IPRST_3_SET;
+    }
+    while (iprst3 != SRST_IPRST_3_SET);
```
This is CPU intensive. Could there be a delay with typical reset time.

0xc0170 replied:
As this is 3rd party driver (SDK in this case), not a blocker.

What is outstanding in this PR

ganesh-ramachandran replied:
> `This is CPU intensive. Could there be a delay with typical reset time.`

@0xc0170 @mazimkhan, addressed this comment, with the updated driver received from Toshiba Japan.

cmonr commented:
@hanno-arm @mazimkhan It looks like this PR is good to start CI. Let us know if that isn't the case.

cmonr commented:
Changes were just a rebase. Starting CI.

/morph build

cmonr commented:
Will still need a final OK from @hanno-arm and @mazimkhan to bring this in assuming all is good.

ganesh-ramachandran commented:
@cmonr, CI build issue - case sensitive header inclusion resolved, and rebased.

0xc0170 commented:
/morph build

hanno-arm commented:
@mazimkhan Please review.

0xc0170 commented:
/morph export-build

cmonr commented:
Looks like NUCLEO_F429ZI boards caused test to time out.

Relaunching.
/morph test

cmonr commented:
/morph mbed2-build

0xc0170 commented:
/morph mbed2-build

0xc0170 commented on targets/TARGET_TOSHIBA/TARGET_TMPM46B/flash_api.c:
```diff
@@ -0,0 +1,163 @@
+    }
+
+    core_util_critical_section_exit();
+    return status;
```
ERROR is 1, but this API if an error is, there should be -1. Ca nyou pleas elook at HAL return types if they match?

ganesh-ramachandran replied:
@0xc0170, Yes, on ERROR -1 to be returned. Added FAIL with -1 and updated.

0xc0170 commented on targets/TARGET_TOSHIBA/TARGET_TMPM46B/spi_api.c:
```diff
@@ -0,0 +1,295 @@
+    SPI_SPH0_SPO1,       // [Clock Phase 0, Clock Polarity 1] = 1
+    SPI_SPH1_SPO0,       // [Clock Phase 1, Clock Polarity 0] = 2
+    SPI_SPH1_SPO1        // [Clock Phase 1, Clock Polarity 1] = 3
+} SPI_MODE;
```
this is confusing, one enum is `SPI_MODE` and right below `spi_mode`.

ganesh-ramachandran replied:
@0xc0170, removed these enums.

0xc0170 commented on targets/TARGET_TOSHIBA/TARGET_TMPM46B/sleep.c:
```diff
@@ -0,0 +1,33 @@
+    // deepsleep = sleep because, TMPM46BF10FG does not support the low power
+    // consumption mode configured with the SLEEPDEEP bit in the Cortex-M4 core.
+    // Setting the bit of the system control register is prohibited.
+    hal_sleep();
```
if this is the case, then does this mean, deep sleep would not turn of high freq clocks in this case (would it need to be manual here or ?) 

cc @c1728p9

0xc0170 replied:
@ganesh-ramachandran How many low power modes does this MCU have? Is it just one ? No more further low power oriented ? Like some clocks off, only few peripherals active, etc?

ganesh-ramachandran replied:
@0xc0170, Implemented deepsleep using the below information:
TMPM46BF10FG has three low power consumption modes: IDLE, STOP1 and STOP2. To shift to the low power consumption mode, specify the mode in the system control register and execute the WFI instruction. In this case, execute reset or generate the interrupt to release the mode.
Note 1: The TMPM46BF10FG does not offer any event for releasing the low power consumption mode. Transition to the low power consumption mode by executing the WFE instruction is prohibited.
Note 2: The TMPM46BF10FG does not support the low power consumption mode configured with the SLEEPDEEP bit in the Cortex-M4 core. Setting the <SLEEPDEEP> bit of the system control register is prohibited.
Note 3: Transition to the power consumption mode while the warm-up timer is operating is prohibited.

ganesh-ramachandran replied:
@0xc0170, Please review and confirm the deepsleep changes.

0xc0170 commented on targets/TARGET_TOSHIBA/TARGET_TMPM46B/us_ticker.c:
```diff
@@ -0,0 +1,138 @@
+}
+
+// initialize us_ticker
+void us_ticker_init(void)
```
I've noticed in the init interrupts for HAL. you are not using NVIC_SetVector - dynamic vectors?

See https://docs.mbed.com/docs/mbed-os-handbook/en/5.2/advanced/porting_guide/ (Bootstrap section)

ganesh-ramachandran replied:
@0xc0170, Added NVIC_SetVector - dynamic vectors

0xc0170 commented on targets/TARGET_TOSHIBA/TARGET_TMPM46B/us_ticker.c:
```diff
@@ -0,0 +1,138 @@
+const ticker_info_t* us_ticker_get_info()
+{
+    static const ticker_info_t info = {
+			1875000,//1875000,
```
are there tabs as this line is misaligned with line below?

0xc0170 commented on targets/TARGET_TOSHIBA/TARGET_TMPM46B/trng_api.c:
```diff
@@ -0,0 +1,97 @@
+    // Interrupt clearing
+    ret = ESG_ClrInt();
+    if (ret == ERROR)
+    {
```
`{` one the same line (line 74 as well here)

cmonr commented:
Still waiting on @c1728p9 for feedback on @0xc0170 comment.

0xc0170 commented:
> Still waiting on @c1728p9 for feedback on @0xc0170 comment.

Should not block this? It's more for the ongoing specification in sleep.

0xc0170 commented:
/morph build

mazimkhan commented:
@ganesh-ramachandran thanks for incorporating the comments. Approved.
